[WIP] [FC-0036] Search Content Tags

yusuf-musleh opened the pull request:
## Description

This PR adds the the ability to search tags in the Content Tags Drawer.

![Screen Shot 2023-12-04 at 1 49 05 PM](https://github.com/open-craft/frontend-app-course-authoring/assets/6829768/0cbb6077-042d-47e2-9e22-fe64de075ca1)

## Supporting information

Related Tickets:
- https://github.com/openedx/modular-learning/issues/122
- Depends on https://github.com/openedx/frontend-app-course-authoring/pull/704

## Testing instructions

1. Run you local dev stack.
2. Next, navigate to http://localhost:18010/admin/waffle/flag/ and create a new flag: new_studio_mfe.use_tagging_taxonomy_list_page make sure it is active for everyone
3. If you don't already have sample taxonomy/tags data, follow the instructions in this repo to generate sample data: https://github.com/open-craft/taxonomy-sample-data
4. Navigate to the sample courses and open the tags drawer for the units
5. Check that all the taxonomies are shown in the drawer and the applied tags are shown
6. Expand some taxonomies and click on the "Add Tags" button
7. In the search bar, type in a term to search for in the tags
8. Confirm that the search results contain the search term. If the root tags shown do not contain the search term, that means the term is in one its (grand)children
9. Confirm that when you check/uncheck a tag it is reflected in the tags tree
10. Confirm that if you check a tag, all its ancestors become implicit
11. Confirm when you uncheck a tag and it is the only child, all its ancestors get unchecked
12. Confirm that the tags number gets updated correctly whenever you add/remove tags
---
Private-ref: [FAL-3572](https://tasks.opencraft.com/browse/FAL-3572)

yusuf-musleh committed:
feat: Add tag search in content tags drawer

yusuf-musleh commented on src/content-tags-drawer/ContentTagsDropDownSelector.jsx:
```diff
@@ -25,25 +28,32 @@ const ContentTagsDropDownSelector = ({
-  // use useQueries to load all the pages based on that.
-  const [fetchUrl, setFetchUrl] = useState(subTagsUrl);
+  const [currentPage, setCurrentPage] = useState(1);
+
+  useEffect(() => {
+    if (searchTerm !== prevSearchTerm.current) {
+      setCurrentPage(1);
+    }
+  }, [searchTerm]);
 
   const isOpen = (i) => dropdownStates[i];
+  const closeAllDropdowns = () => {
+    const updatedStates = { ...dropdownStates };
+    // eslint-disable-next-line no-return-assign
+    Object.keys(updatedStates).map((key) => updatedStates[key] = false);
+    setDropdownStates(updatedStates);
+  };
 
   const clickAndEnterHandler = (i) => {
     // This flips the state of the dropdown at index false (closed) -> true (open)
     // and vice versa. Initially they are undefined which is falsy.
     setDropdownStates({ ...dropdownStates, [i]: !dropdownStates[i] });
   };
 
-  const { data: taxonomyTagsData, isSuccess: isTaxonomyTagsLoaded } = useTaxonomyTagsData(taxonomyId, fetchUrl);
+  const {
+    data: taxonomyTagsData,
+    isSuccess: isTaxonomyTagsLoaded,
+  } = useTaxonomyTagsData(taxonomyId, subTagsUrl, currentPage, searchTerm);
 
```
@bradenmacdonald I'm trying to reset the `currentPage` state to 1 whenever the `searchTerm` changes. However the `useTaxonomyTagsData` is firing twice since it seems like the state is changing after the new `searchTerm` prop is passed in (and the hook already fires). Any suggestions on how I can make them tightly coupled, i.e. whenever the `searchTerm` changes the `currentPage` should reset before the hook fires?

bradenmacdonald replied:
@yusuf-musleh 
First of all, I think you generally are using `useEffect` too much - take a quick read of [You Might Not Need an Effect](https://react.dev/learn/you-might-not-need-an-effect#):

> If there is no external system involved (for example, if you want to update a component’s state when some props or state change), you shouldn’t need an Effect

So `useEffect` is not the right approach here.

This exact situation is explained in [Adjusting some state when a prop changes](https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes) so give that a read too. Although their primary recommendation would work but still may result in an extra fetch. So here's a way to avoid that:

### Complex solution

```
const [currentPageState, setCurrentPage] = useState(1);
const [currentPageFor, setCurrentPageFor] = useState(searchTerm);
let currentPage = currentPageState;
if (searchTerm !== currentPageFor) {
    // When the search term changes, reset to page 1
    setCurrentPage(1);
    setCurrentPageFor(searchTerm);
    // Apply this immediately to avoid a wasted fetch of later pages with the new search term
    currrentPage = 1;
}
```

### Simple solution

See [Resetting all state when a prop changes](https://react.dev/learn/you-might-not-need-an-effect#resetting-all-state-when-a-prop-changes).

yusuf-musleh replied:
@bradenmacdonald Thanks for the pointers and the reference! It seems like I misunderstood how useEffect should be used, and stared spamming it everywhere 😂  I went with a flavor of the complex approach, since the simple approach was reseting the whole dropdown, whenever you'd select a tag. It's working quite well now, not firing/rendering multiple times when entering a search term:

```js
  const [currentPage, setCurrentPage] = useState(1);
  const [prevSearchTerm, setPrevSearchTerm] = useState(searchTerm);

  // Reset the page and tags state when search term changes
  // and store search term to compare
  if (prevSearchTerm !== searchTerm) {
    setPrevSearchTerm(searchTerm);
    setCurrentPage(1);
    closeAllDropdowns();
    setTags([]);
  }
```

yusuf-musleh committed:
style: Fix style/spacing inconsistencies

yusuf-musleh committed:
fix: Spacing of tags in dropdown selector

yusuf-musleh committed:
feat: Implement pre-loading tags

This adds support for pre-loading taxonomy tags if they fit within the
threshold to avoid making exessive requests, while still maintaining
support for loading more paginated results.

yusuf-musleh committed:
refactor: Remove not needed encoding

yusuf-musleh committed:
feat: Add content tags tree state + editing (#704)

This commit adds the add/remove functionality of content tags where the
state is stored and changes are updated in the backend through the API.
Changes are reflected in the UI automatically.

KristinAoki committed:
fix: wrong min count alert showing (#730)

KristinAoki committed:
fix: grading segment number ranges (#729)

yusuf-musleh committed:
test: Update/fix tests

yusuf-musleh commented:
Closing in favor of https://github.com/openedx/frontend-app-course-authoring/pull/737 to merge to main repo since the base of this PR cannot be changed to main repo. cc: @bradenmacdonald
